### DISCUSSION TITLE
Streaming database results instead of buffering in API

### DISCUSSION THREAD

**Discussion by daawaan4x** [directus/directus#16230] [Feature Requests]
What would be the feasibility of implementing knex streams being piped directly into express response handlers instead of buffering it to lessen memory load in the API? Although it requires an overhaul of how results are passed around in the API since afaik all internal services/controllers buffer query results, fortunately this can be incrementally adopted by creating `[*]stream` functions side-by-side with the preexisting buffering ones.

**Comment by rijkvanzanten** [comment]
Good idea! I haven't done too much research on this subject yet, but here's a couple immediate thoughts on difficulty: 

1. Not every database supports streams, so we'll have to come up with some good handler for both and/or an abstraction layer that will wrap a memory object into a readable steam so we can keep using the same handlers. I know knex already handles some of that, but I'd like to take the time researching a level above that, so we can support other non-sql sources down the line too 

2. Similarly, we often need to transform multi-level data sets (for example to stitch together nested results), which feels like it'll be harder to do efficiently if we don't have full sets yet. (For example we'd have to callback on every item in the stream and push it to the right nested array, instead of making the nested array a .filter of the known parents). Curious to see how that balances out performance wise. Might be faster still :) 

3. I'm curious about querying performance at smaller data sizes. I believe some connection libs use cursors and other tactics to add streaming. No idea how that overhead affects querying speed (if at all)

**Reply by daawaan4x** [reply]
@rijkvanzanten A quick thought about point 2, I think a number of the databases that are currently being supported have native json aggregation functions so maybe all the transformations could be offloaded to the database instead resulting into a single query only for nested results. Haven't really checked all the docs for the supported databases but i'm curious about the possible hindrances and penalties to this approach. AFAIK hasura does this in compiling their GraphQL ASTs to SQL queries but falls back to multi-queries instead in case the database doesn't natively support json aggregation

**Reply by rijkvanzanten** [reply]
> I think a number of the databases that are currently being supported have native json aggregation functions so maybe all the transformations could be offloaded to the database instead resulting into a single query only for nested results

Yeah we've tried that in the past, but there's one big gotcha: we need to be able to limit the amount of nested result _per parent_, so "get 50 root items, and nest 3 o2m items per root item". There's currently no real way in SQL to do that nested LIMIT. From what I've seen so far, the most efficient way to get that is by doing separate batched queries with `parent_id WHERE IN (x, y, z)` and stitching the results after the fact on the app layer side of things 🤔

**Reply by rijkvanzanten** [reply]
(that only applies to one to many by the way, not many to one 🙂

**Reply by daawaan4x** [reply]
Just a speculation, haven't tried anything yet, wouldn't joins mitigate the limit issue though?

**Reply by rijkvanzanten** [reply]
> Just a speculation, haven't tried anything yet, wouldn't joins mitigate the limit issue though?

You could if you join on a sub query selection, but at that point you're doing a query per parent row anyways, at which point it'd be faster to run them in parallel (AFAIK)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
